### PR TITLE
BC break: Calling a method unnecessarily

### PR DESCRIPTION
**Pull request by thiagotalma**
Making an indirect call to the `$this->getOldAttributes()` method caused my code to break. I understand that the correct thing would be to reference `$this->_oldAttributes`.

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #19272

### CODE REVIEW

File: framework/db/BaseActiveRecord.php
Comment (bizley) [inline framework/db/BaseActiveRecord.php:1766]: If I read this correctly here might be a problem when `$this->_oldAttributes` is `null`.

**Comment by schmunk42** [comment]
> caused my code to break. 

What broke? Were there error messages, wrong data?

File: framework/db/BaseActiveRecord.php
Comment (thiagotalma) [reply framework/db/BaseActiveRecord.php:1766]: https://github.com/yiisoft/yii2/blob/2a7c6f2ca1b362a26ef580a280c188dda2a74510/framework/db/BaseActiveRecord.php#L642

In theory, this check ensures that this problem will never occur.

**Comment by thiagotalma** [comment]
The problem was because I was overriding the `_get()` method.

I only opened the PR because it may be that the same problem occurs with other people. In my understanding, it makes no sense to force the call in the `$this->getOldAttributes` method without the slightest need, since the `$this->_oldAttributes` is available.

Anyway, **MY** problem I already solved.
Feel free to close the PR. Or improve it, because, in my opinion, that wasn't the most adequate implementation. I'm open to suggestions.

**Comment by bizley** [comment]
I think we could change it to use private property indeed but with additional check for null.

**Comment by thiagotalma** [comment]
@bizley done

**Review by bizley — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

**Comment by samdark** [comment]
Would you please a line for CHANGELOG?

**Comment by thiagotalma** [comment]
@samdark done!

**Comment by bizley** [comment]
Sorry, just noticed that this is wrong, let's make it like that for example

```
private function isValueDifferent($newValue, $oldValue)
{
    if (is_array($newValue) && is_array($oldValue)) {
        $newValue = ArrayHelper::recursiveSort($newValue);
        $oldValue = ArrayHelper::recursiveSort($oldValue);
    }

    return $newValue !== $oldValue;
}
```

then we can just call it like

```
if (isset($names[$name]) && (!array_key_exists($name, $this->_oldAttributes) || $this->isValueDifferent($value, $this->_oldAttributes[$name]))) {
    $attributes[$name] = $value;
}
```

Otherwise we compare $value to an empty array if not set in old attributes.

**Review by bizley — CHANGES_REQUESTED** [review]
Please modify as per my comment.

**Comment by samdark** [comment]
Thanks.

**Comment by samdark** [comment]
Owww... GitHub did not update the comments before I hit the merge button :(

**Comment by samdark** [comment]
@thiagotalma would you please fix it as @bizley suggested?

**Pull request merged**
